src/java/org/apache/cassandra/config/registry/TypeConverterRegistry.java

```diff
@@ -0,0 +1,131 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.cassandra.config.registry;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Objects;
+
+import org.apache.cassandra.config.DataRateSpec;
+import org.apache.cassandra.config.DataStorageSpec;
+import org.apache.cassandra.config.DurationSpec;
+import org.apache.cassandra.db.ConsistencyLevel;
+import org.apache.cassandra.exceptions.ConfigurationException;
+
+import static org.apache.commons.lang3.ClassUtils.primitiveToWrapper;
+
+/**
+ * A registry for {@link TypeConverter} instances.
+ */
+public class TypeConverterRegistry
+{
+    public static final TypeConverterRegistry instance = new TypeConverterRegistry();
+    private final Map<ConverterKey, TypeConverter<?>> converters = new HashMap<>();
+
+    private TypeConverterRegistry()
+    {
+        registerConverters(converters);
+    }
+
+    public <V> TypeConverter<V> get(Class<?> from, Class<V> to)
+    {
+        Class<?> fromShaded = primitiveToWrapper(from);
+        TypeConverter<V> converter = get(fromShaded, to, null);
+        if (converter == null)
+            throw new ConfigurationException(String.format("No converter found from '%s' to '%s'", fromShaded.getCanonicalName(), to.getCanonicalName()));
+        return converter;
+    }
+
+    @SuppressWarnings("unchecked")
+    public <V> TypeConverter<V> get(Class<?> from, Class<V> to, TypeConverter<V> defaultConverter)
+    {
+        Class<?> fromShaded = primitiveToWrapper(from);
+        if (fromShaded.equals(to))
+            return to::cast;
+        if (converters.get(key(fromShaded, to)) == null)
+            return defaultConverter;
+        return (TypeConverter<V>) converters.get(key(fromShaded, to));
+    }
+
+    private static void registerConverters(Map<ConverterKey, TypeConverter<?>> converters)
```

Comment:
we have `org.apache.cassandra.config.YamlConfigurationLoader#updateFromMap` which makes sure we always reuse the same input set that was supported in YAML... is there a reason to reinvent this?